oclc/classify.py

```diff
@@ -26,15 +29,204 @@
 )
 from core.util import MetadataSimilarity
 from core.util.xmlparser import XMLParser
+from coverage_utils import MetadataWranglerBibliographicCoverageProvider
 from viaf import NameParser as VIAFNameParser
 
-
 class OCLC(object):
     """Repository for OCLC-related constants."""
     EDITION_COUNT = "OCLC.editionCount"
     HOLDING_COUNT = "OCLC.holdings"
     FORMAT = "OCLC.format"
 
+class OCLCClassifyXMLParser(XMLParser):
+    """Turn the output of the OCLC Classify API into a Metadata object.
+
+    This object includes ContributorData for the book's contributors,
+    ClassificationData representing the most common classifications for
+    the book, and MeasurementData describing how many libraries have
+    collected the book.
+    """
+
+    # OCLC in-representation 'status codes'
+    SINGLE_WORK_SUMMARY_STATUS = 0
+    SINGLE_WORK_DETAIL_STATUS = 2
+    MULTI_WORK_STATUS = 4
+    NO_INPUT_STATUS = 100
+    INVALID_INPUT_STATUS = 101
+    NOT_FOUND_STATUS = 102
+    UNEXPECTED_ERROR_STATUS = 200
+
+    INTS = set([OCLC.HOLDING_COUNT, OCLC.EDITION_COUNT])
+
+    NAMESPACES = {'oclc' : 'http://classify.oclc.org'}
+    ROLES = re.compile("\[([^]]+)\]$")
+    LIFESPAN = re.compile("([0-9]+)-([0-9]*)[.;]?$")
+
+    CLASSIFIERS = {"fast": Subject.FAST, "lcc": Subject.LCC, "ddc": Subject.DDC}
+
+    @classmethod
+    def parse(cls, _db, xml):
+        tree = etree.fromstring(xml, parser=etree.XMLParser(recover=True))
+        contributors = cls.contributors(_db, tree)
+        measurements = cls.measurements(_db, tree)
+        subjects = cls.subjects(_db, tree)
+
+        return Metadata(
+            data_source=DataSource.OCLC,
+            contributors=contributors,
+            measurements=measurements,
+            subjects=subjects,
+        )
+
+
+    @classmethod
+    def get_work_tags(cls, tree):
+        return cls._xpath(tree, "//oclc:work")
+
+    # CONTRIBUTORS:
+
+    @classmethod
+    def contributors(cls, _db, tree):
+        """Returns a list of ContributorData objects"""
+        # We prefer to get author information out of the <authors> tag,
+        # but sometimes all we have is a <work> tag.
+        return cls.get_authors(tree) or cls.get_authors_from_work_tag(_db, tree)
+
+    @classmethod
+    def get_authors_from_work_tag(cls, _db, tree):
+        # Sometimes, different versions of the same book get grouped together
+        # as individual <work> tags under a <works> list.
+        author_lists = []
+        work_tags = cls.get_work_tags(tree)
+        if work_tags is not None:
+            for work_tag in work_tags:
+                author_string = work_tag.get('author')
+                if author_string:
+                    author_lists.append(NameParser.parse_multiple(author_string))
+
+        # If there are multiple lists of authors (as a result of multiple <work> tags),
+        # use the shortest list.
+        return sorted(author_lists, key=len)[0]
```

Comment:
Here's an example of a multi-work response:

```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<classify xmlns="http://classify.oclc.org">
  <response code="4"/>
  <!--Classify is a product of OCLC Online Computer Library Center: http://classify.oclc.org-->
  <workCount>2</workCount>
  <start>0</start>
  <maxRecs>25</maxRecs>
  <orderBy>thold desc</orderBy>
  <input type="isbn">9788888211756</input>
  <works>
    <work author="Manzoni, Alessandro, 1785-1873" editions="30" format="eAudiobook" holdings="59" hyr="2015" itemtype="itemtype-audiobook-digital" lyr="1900" owi="1151523604" schemes="DDC LCC" title="I promessi sposi" wi="123476150"/>
    <work author="Manzoni, Alessandro, 1785-1873" editions="1" format="Book" holdings="1" hyr="2009" itemtype="itemtype-book" lyr="2009" owi="1154035336" schemes="DDC" title="I promessi sposi : audiolibro integrale" wi="555641639"/>
  </works>
</classify>
```

Here's my thinking:

1. The #1 thing we need out of this is the author's display name, which in this case is "Alessandro Manzoni". The next most important things are the popularity and subject information.
2. To have a decent chance of knowing the author's display name (as opposed to guessing it), we need the author's VIAF ID or LOC ID.
3. We have some popularity information here but nothing else -- no VIAF ID or LOC ID, no subject information.
4. We can't get that information from this document, but we can get it by doing a _second_ lookup based on one of the OWI (OCLC Work IDs) contained in this document.
5. So the best thing for the parser to do is to assemble a list of two Metadata objects, one for each of the `<work>` tags.
6. The author information (obtained through `NameParser.parse_multiple`) can go into those Metadata objects, but the really important thing is the `IdentifierData` containing the OWI (type=Identifier.OCLC_WORK).

So that's the case where `parse` returns multiple results. If we get that going I think we'll be done with the parser part; then it'll just be a case of improving the IdentifierLookupCoverageProvider to handle all these cases.